### DISCUSSION TITLE
Cannot return null for non-nullable field

### DISCUSSION THREAD

**Discussion by theprantadutta** [mikro-orm/mikro-orm#2465] [Q&A]
**Describe the bug**
I am trying to make a Graphql server with Mikro-ORM. Basically, I have a one-to-many relationship with category and product. Each product should belong to a category and a category can have multiple products. Now I am trying to fetch all the products with their categories and I am having the following error.

**Stack trace**
```
    "message": "Cannot return null for non-nullable field Category.title.",
      "locations": [
        {
          "line": 11,
          "column": 7
        }
      ],
      "path": [
        "products",
        0,
        "category",
        "title"
      ],
      "extensions": {
        "code": "INTERNAL_SERVER_ERROR",
        "exception": {
          "stacktrace": [
            "Error: Cannot return null for non-nullable field Category.title.",
...
```

**To Reproduce**
1. My Product.ts file
```
      import {
      Entity,
      IdentifiedReference,
      ManyToOne,
      PrimaryKey,
      Property
    } from '@mikro-orm/core'
    import { Field, Float, ObjectType } from 'type-graphql'
    import { ProductArguments } from '../util/types'
    import { Category } from './Category'
    
    @Entity()
    @ObjectType()
    export class Product {
      constructor({ title, price, description }: ProductArguments) {
        this.title = title
        this.price = price
        this.description = description
      }
    
      @Field()
      @PrimaryKey()
      id!: number
    
      @Field()
      @Property()
      title!: string
    
      @Field()
      @Property()
      imageUrl!: string
    
      @Field(() => Float)
      @Property({ type: 'decimal' })
      price!: number
    
      @Field(() => String)
      @Property({ type: 'text' })
      description!: string
    
      @Field(() => Category)
      @ManyToOne(() => Category)
      category!: IdentifiedReference<Category>
    
      @Field()
      @Property()
      createdAt: Date = new Date()
    
      @Field()
      @Property({ onUpdate: () => new Date() })
      updatedAt: Date = new Date()
    }
```
2. My Category.ts file
```
    import {
    Cascade,
      Collection,
      Entity,
      LoadStrategy,
      OneToMany,
      PrimaryKey,
      Property
    } from '@mikro-orm/core'
    import { Field, ObjectType } from 'type-graphql'
    import { Product } from './Product'
    
    @Entity()
    @ObjectType()
    export class Category {
      constructor(title: string) {
        this.title = title
      }
    
      @Field()
      @PrimaryKey()
      id!: number
    
      @Field()
      @Property()
      title!: string
    
      @Field(() => [Product])
      @OneToMany({
        entity: () => Product,
        mappedBy: (product) => product.category,
        cascade: [Cascade.REMOVE],
        strategy: LoadStrategy.JOINED
      })
      products = new Collection<Product>(this)
    
      @Field()
      @Property()
      createdAt: Date = new Date()
    
      @Field()
      @Property({ onUpdate: () => new Date() })
      updatedAt: Date = new Date()
    }
```
3. Now in my resolvers, I am trying to get all the categories with the following.
```
   const categories = await em
      .createQueryBuilder(Category)
      .select('*')
      .getResult()
    return em.populate(categories, ['products'])
```
This works perfectly fine. No errors.
But when I try to get all products with their categories like the following:
```
   const products = await em
      .createQueryBuilder(Product)
      .select('*')
      .getResult()
    return em.populate(products, ['category'])
```
Or with left join
```
   return em
      .createQueryBuilder(Product, 'p')
      .select('*')
      .leftJoinAndSelect('p.category', 'c')
      .getResultList()
```
I got the error described above.
Here's the generated query.
>> select "p".*, "c"."id" as "c__id", "c"."title" as "c__title", "c"."created_at" as "c__created_at", "c"."updated_at" as "c__updated_at" from "product" as "p" left join "category" as "c" on "p"."category_id" = "c"."id"```

I also tried to run this query in the database and I got the actual result that I wanted. But in GraphQL Server, I got the above error. What am I doing wrong here?

**Expected behavior**
I want it return my products like the following: 
```
data: {
  products: [
    {
      id: 1,
      title: 'T-shirts',
      ...other staff,
      categories: {
        id
        title
        ...other_stuff
      }
    }
  ]
}
```

**Additional context**
You can check out my code for getting all categories too. That's working fine. But I can't get my products. 

**Versions**

| Dependency | Version |
| - | - |
| node | 16.1 |
| typescript | 4.3 |
| mikro-orm | next(5.x) |
| your-driver | postgresql |

**Comment by B4nan** [comment]
This is not comming from MikroORM, if you want to report bugs in here, please don't use GQL for reproductions so it is easier to understand. 

Looking at your code, I found one issue in the entity definition. If you do not use ts-morph metadata provider, you need to adjust this one:

```ts
      @Field(() => Category)
      @ManyToOne(() => Category, { wrappedReference: true }) // added wrappedReference 
      category!: IdentifiedReference<Category>
```

As with the default metadata provider this information needs to be passed explicitly. 

Other ideas:
- ensure you actually have the column as non nullable and it has the data for rows you select
- try logging what the `qb.getResult()` gives you (maybe after you call `em.populate()`
- if you don't have any specific reason to use QB, I would rather use EM directly, it's less verbose

**Reply by theprantadutta** [reply]
@B4nan Thank you for your help.  Sorry, if this was a noob mistake. But I just started working on mikro-orm. Anyway, can you provide the query with EM to get all the products with categories? I can't find the proper query in the documentation. The query I found, it always executes two SQL queries. And I am using `metadataProvider: TsMorphMetadataProvider` in my mikro-orm.config.ts file. Also, I console logged `qb.results()`, I got all the products and the category actually contains `Reference { entity: [Ref<Category>] }` . 

This is a mikro-orm problem. The problem here is that I am using `IdentifiedReference` for my Category model with my Product model which actually provides some map methods to get the property like `getProperty` and others. When I removed the `IdentifiedReference<Category>` with `Category` , it works fine. What I didn't understand is that I can't use `IdentifiedReference` to get the data in graphql.

Anyway, can you provide the actual em query for getting all the products with categories using a single SQL query? I am getting the result I want with em, but it executes two sql query. When I use queryBuilder, I get the result I want using a single SQL query. I can't do it with EM. Thanks btw.

**Reply by B4nan** [reply]
> I got all the products and the category actually contains Reference { entity: [Ref<Category>] } .

This means the category is not populated, otherwise it would say `Reference { entity: [Category] }` without the `Ref<...>` - so it won't be part of serialized response (only it's PK, not as an object).

> This is a mikro-orm problem. 

Then provide MikroORM reproduction, without GQL. Reproductions need to be minimal. 

> When I removed the IdentifiedReference<Category> with Category , it works fine. 

That sounds exactly as if you would be missing the `wrappedReference: true`. Have you actually tried to put it there? 

Also be sure to update to latest version. I see you said you are using `next(5.x)`, but the whole point of the versions in issue template is to say the exact version you are on, not latest or next, or a version range, that is all wrong - it should be things like `5.0.0-dev.123`. Things are moving every day :] I was actually fixing something with identified references few days ago.

> Anyway, can you provide the actual em query for getting all the products with categories?

```ts
// this:
const products = await em
    .createQueryBuilder(Product)
    .select('*')
    .getResult()
return em.populate(products, ['category'])

// is the same as this:
const products = await em.find(Product, {}, {
  populate: ['category'],
})
```

**Reply by theprantadutta** [reply]
Everything else is fine. But 
> const products = await em.find(Product, {}, {
  populate: ['category'],
})

I actually did do this. But it executes two queries. But I want it to generate only one. That's the reason I even considered doing this with the querybuilder.

**Reply by B4nan** [reply]
```
const products = await em.find(Product, {}, {
  populate: ['category'],
  strategy: LoadStrategy.JOINED,
})
```

**Reply by theprantadutta** [reply]
@B4nan thanks a lot, that's exactly what I wanted.
